src/include/txn_inline.h

```diff
@@ -489,22 +558,109 @@ __txn_should_assign_timestamp(WT_SESSION_IMPL *session, WT_TXN_OP *op)
     return (true);
 }
 
+/*
+ * __wt_txn_timestamp_usage_check --
+ *     Check if a commit will violate timestamp rules.
+ */
+static WT_INLINE int
+__wt_txn_timestamp_usage_check(
+  WT_SESSION_IMPL *session, WT_TXN_OP *op, wt_timestamp_t op_ts, wt_timestamp_t prev_op_durable_ts)
+{
+    WT_BTREE *btree;
+    WT_TXN *txn;
+    uint16_t flags;
+    char ts_string[2][WT_TS_INT_STRING_SIZE];
+    const char *name;
+    bool no_ts_ok, txn_has_ts;
+
+    btree = op->btree;
+    txn = session->txn;
+    flags = btree->dhandle->ts_flags;
+    name = btree->dhandle->name;
+    txn_has_ts = F_ISSET(txn, WT_TXN_HAS_TS_COMMIT | WT_TXN_HAS_TS_DURABLE);
+
+    /* Timestamps are ignored on logged files. */
+    if (F_ISSET(btree, WT_BTREE_LOGGED))
+        return (0);
+
+    /*
+     * Do not check for timestamp usage in recovery. We don't expect recovery to be using timestamps
+     * when applying commits, and it is possible that timestamps may be out-of-order in log replay.
+     */
+    if (F_ISSET(S2C(session), WT_CONN_RECOVERING))
+        return (0);
+
+    /* Check for disallowed timestamps. */
+    if (LF_ISSET(WT_DHANDLE_TS_NEVER)) {
+        if (!txn_has_ts)
+            return (0);
+
+        __wt_err(session, EINVAL,
+          "%s: " WT_TS_VERBOSE_PREFIX "timestamp %s set when disallowed by table configuration",
+          name, __wt_timestamp_to_string(op_ts, ts_string[0]));
+#ifdef HAVE_DIAGNOSTIC
```

Comment:
Can we use WT_ASSERT here instead?

Reply:
Assert also we make the system fail, so either way we are aborting the execution.